torch/ao/quantization/dynamic_tracing/quantization_state.py

```diff
@@ -0,0 +1,725 @@
+        processing is complete.
+        """
+        if op_needs_quantization(cur_op):
+            self.idx += 1
+
+    def _outputs_prepare_hook_tensor(
+        self,
+        output: torch.Tensor,
+        first_call: bool,
+        qtensor_id: List[int],
+    ) -> torch.Tensor:
+        if first_call:
+            if not hasattr(output, '_qtensor_info'):
+                # TODO: use actual dtype instead of defaulting to float
+                output._qtensor_info = QTensorInfo(qtensor_id[0], torch.float)  # type: ignore[attr-defined]
+                qtensor_id[0] += 1
+            self.output_qtensor_infos.append(output._qtensor_info)  # type: ignore[attr-defined]
+            # TODO(future PR): add an observer if needed
+        return output
+
+    def outputs_prepare_hook(
+        self,
+        outputs: Any,
+        first_call: bool,
+        qtensor_id: List[int],
+    ) -> Any:
+        """
+        This function is expected to be called on the outputs of a prepared
+        module right before they are returned to the parent.
+        """
+        # TODO: handle objects with deeper nested tensors
+        if first_call:
+            if isinstance(outputs, torch.Tensor):
+                self._outputs_prepare_hook_tensor(outputs, first_call, qtensor_id)
+            elif isinstance(outputs, tuple):
+                # TODO: handle other tuple subclasses more generically
+                new_outputs = []
+                for output in outputs:
+                    if isinstance(output, torch.Tensor):
+                        new_outputs.append(self._outputs_prepare_hook_tensor(
+                            output, first_call, qtensor_id))
+                    else:
+                        new_outputs.append(output)
+                # hacky check for collections.namedtuple, TODO improve this
+                # https://stackoverflow.com/questions/2166818/how-to-check-if-an-object-is-an-instance-of-a-namedtuple
+                if hasattr(outputs, '_fields'):
+                    return outputs.__class__(*new_outputs)
+                else:
+                    return tuple(new_outputs)
+            else:
+                pass
+
+        return outputs
+
+    def outputs_convert_hook(
+        self,
+        outputs: Any,
+    ) -> Any:
+        """
+        This function is expected to be called on the outputs of a converted
+        module right before they are returned to the parent.
+        """
+
+        if isinstance(outputs, torch.Tensor):
+            qtensor_info = self.output_qtensor_infos[0]
+            if self.output_dtypes is not None:
+                assert qtensor_info is not None
+                # check the output dtype, and do the conversion if needed
+                output_dtype = self.output_dtypes[0]
+                if qtensor_info.inf_dtype != output_dtype:
+                    assert output_dtype is torch.float, \
+                        'non-float output dtypes not handled yet'
```

Comment:
nit: might also be good to add an assert for the `inf_dtype`, currently we have `quint8, qint8 and float16`